### ISSUE TITLE
Interoperable handling of invalid markup 

### ISSUE THREAD

**Issue by fred-wang** [mathml-refresh/mathml#15]
https://mathml-refresh.github.io/mathml/chapter2.html#interf.error

> If a MathML-input-conformant application receives input containing one or more elements with an illegal number or type of attributes or child schemata, it should nonetheless attempt to render all the input in an intelligible way, i.e., to render normally those parts of the input that were valid, and to render error messages (rendered as if enclosed in an merror element) in place of invalid expressions.

Gecko renders (not very useful) "invalid-markup" message
WebKit just layout a 0x0 box for the invalid subtree

I think the latter is more efficient and less code so I'd prefer it. Probably if we want to log errors, it should be in the console instead.

**Comment by fred-wang** [comment]
Adding mathml4 label too, as we might want to adjust the  behavior.

**Comment by fred-wang** [comment]
cc @rwlbuis

**Comment by fred-wang** [comment]
Just for the record, there is a third option which is probably a bit more complicate to implement/describe but more aligned with to what people expect with CSS / HTML5: Add anonymous `<mrow>` elements (e.g. one for numerator and one for denominator for `mfrac`) to ensure that MathML elements have the right number of arguments.

**Comment by fred-wang** [comment]
Quoting @bfgeek 

> See:
> https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=6704
> ... for a simple example.
> 
> https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=6705
> ... for a more complex example.
> 
> The above example indicates that the implementations are trying to do something funny with various inputs which an implementation thinks is invalid.
> 
> I suggest removing these, and instead making sure that all the MathML algorithms can deal with the edge cases which developers might feed them.
> 
> (an alternative would be to strictly define, with algorithms, what type of DOM input is required, and the validation steps, however this would be prone to bugs, and compatibility issues).
> 
> This the a bit of a meta issue, however would need lots of tests with the "edge" cases for all of the different layout algorithms within MathML.

**Comment by fred-wang** [comment]
I'm not sure there is a canonical way to "fix" invalid markup (Ignore extra children? Add empty anonymous? Wrap a subset of children into anonymous containers?). My preference would be to align on WebKit's behavior as it's the easiest to specify/test/implement. This is already how we (at Igalia) implement it our Chromium branch and we would only have to fix Firefox. Basically, my proposal would be:

1) Add a section about invalid markup in MathML Core (maybe import from MathML 4 full), define the invalid constructions (essentially elements that don't have the expected number of children). There is probably an exception do make for mtable, until we clarify what the table layout is.
2) Say that invalid markup has intrinsize size of 0, the layout results as an empty 0x0 box and no painting.
3) In each section where we define algorithm and intrinsic size (#54 etc), say that we do an early return with the result specified in (2)
4) Write testharness tests with all markup of the form `<span style="display: inline-block"><math>INVALID MARKUP</math></span>` as defined in (1) and ensure that the span has size 0 (intrinsic size) and that the "INVALID MARKUP" subtree has size 0x0. Write reftests checking that they don't paint anything (e.g. for fractions, roots...). Probably more tests involving valid and invalid subtrees. I believe we already have tests for invalid markup in Gecko and WebKit that we could "import". However, they would need to be tweaked a bit probably.

Putting this on the agenda of next meeting.

**Comment by bfgeek** [comment]
WebKits implementation causes issues in other areas of the browser. For example the "invalid" <math> is still accessible from VoiceOver, etc.

This type of "easy" fixup produces technical debt in the codebase, and complexity later on which I'd be strongly opposed to adding in chromium. 

Having said that one guiding principle that CSS usually follows is that it should attempt to display all content. This is probably a good principle for MathML to also follow.

I think that for display types which can only accept up to "N" children, inserting any additional children in an anonymous mrow type would be acceptable.

E.g.
<mfrac><mi>a</mi><mi>b</mi><mi>c</mi></mfrac>
would render as:
<mfrac><mi>a</mi><mrow><mi>b</mi><mi>c</mi></mrow></mfrac>

@emilio @heycam

**Comment by bfgeek** [comment]
@tabatkins also might have some good thoughts.

**Comment by fred-wang** [comment]
So the alternative proposal would be:

1. Add a section about invalid markup in MathML Core defining the invalid constructions. Essentially elements they have n children while they expect N != n children. If n < N, they append N - n anonymous empty mrow's. If n > N, they wrap the last N - n + 1 children into an anonymous mrow. An alternative is to always generate N anonymous children but that would lead to bigger render/layout trees and complexity for such elements (even when they are valid) and so is less optimal. Again, there is an exception to make for mtable, were the fixup is instead using anonymous mtr and mtd like in HTML.
2. In each section where we define elements, mention the expected number of children if appropriate (import from MathML 4 full) and link to the general section about how to fix it.
3. Write static reftests checking that invalid markup render the same as their "fixed" version. Also write dynamic versions where n varies between 0 and (at least) N + 2.
4. Implement the behavior for WebKit, Gecko and Chromium.

My main concern (in addition to adding more work) is how dynamic changes will be handled. In the past, WebKit generated some anonymous elements (sometimes two levels of nested anonymous e.g. for scripted elements) in the Renderer/Layout classes and set/remove anonymous style on them: That caused at best inconsistencies after dynamic changes and at worst asserts/crashes. Hopefully, this proposal is a bit safer and we could maybe even handle the Render/Layout tree fixup & reorganization from the DOM classes.

However, I agree that the proposed behavior would be more aligned with HTML5 / CSS (adding the label now btw) and more natural for users, so it's probably worth the added complexity/work.

I don't really like the alternative of "making sure that all the MathML algorithms can deal with the edge cases" as that makes the algorithm more complex and harder to read. For example, the advantage of WebKit's current implementation is that once we know an element is valid we can just call the children and associated variable numerator/denominator/base/superscript etc and that makes the algorithm (which is already relatively complex when we follow all TeX / Open Type MATH metrics) much easier to read and understand.

**Comment by fred-wang** [comment]
> An alternative is to always generate N anonymous children but that would lead to bigger render/layout trees and complexity for such elements (even when they are valid) and so is less optimal.

Since VoiceOver was mentioned above, I however have to say that adding such anonymous children could help accessibility ( #9 ). Typically,

    mfrac
      |_mfrac
      |_mfrac

would become

    mfrac
      |_mrow
        |__frac
      |_mrow
        |__frac

and then we can add roles numerator/denominator on the anonymous mrow's while still keeping roles fraction on the mfrac. Otherwise, this fixup would have to be done in the a11y tree.

**Comment by NSoiffer** [comment]
Doesn't HTML5's [parsing spec](https://html.spec.whatwg.org/multipage/parsing.html) already say what should happen for syntax errors such as missing close tags?

I didn't see in the parsing spec what should happen when the wrong number of arguments are given. Maybe I missed it or maybe because that isn't related to what gets built in the DOM. 

HTML5 has some elements that must be used in ordered pairs (i.e., they don't make sense out of context). For example,  `dl`, `dd`, and `dt` all work together. I saw the parsing spec has some repair mechanisms related to them for parse errors, but I didn't see anything related to inserting missing tags. Does any one what the HTML5 spec says to do in these cases? If there is a precedent (and maybe existing code), it is something we should follow if it makes sense for MathML.

If there is no precedent, then I like the idea of adding anonymous `mrow`s and attaching some CSS to them to visually and hopefully accessibly indicate an error.

**Comment by bfgeek** [comment]
@NSoiffer Yes the parsing spec does specify what should happen with missing close tags, but that isn't what is being discussed.

> I didn't see in the parsing spec what should happen when the wrong number of arguments are given. Maybe I missed it or maybe because that isn't related to what gets built in the DOM.

The primary issue here is that DOM/CSS tries to display all content on the page, HTML doesn't really have a concept of an "incorrect" number of children for any given element (there may be a counter example here, but I'm not aware of any).

E.g. a <table> element can have multiple <caption>s or <thead> even though it only makes "sense" to have one. The browser will display all of them not to drop content.

**Comment by fred-wang** [comment]
An alternative option could be to say that invalid markup are just laid out as mrow. That way we keep a simple fallback similar to the empty 0x0 but ensure that all the content is displayed.

**Comment by davidcarlisle** [comment]
@bfgeek  I think the html spec is trying to specify this

https://html.spec.whatwg.org/#mathml

says

> User agents must act as if any MathML element whose contents does not match the element's content model was replaced, for the purposes of MathML layout and rendering, by a MathML merror element containing some appropriate error message

which means that

`<mfrac><mi>a</mi><mi>b</mi><mi>c</mi></mfrac>`

should render as

`<merror> ...</merror>`

That said, if that proves tricky to specify in an interoperable way, we should change the spec.

**Comment by NSoiffer** [comment]
At the risk of drifting off-topic, is `merror` something that should be in core? It is currently listed in the [core spec](https://mathml-refresh.github.io/mathml-core/#merror), but I don't think that means much at the moment. If it should remain in core, I think we should follow the HTML spec and use it as @davidcarlisle suggested. However, the _HTML5 spec_ is short on detail about what the content of the `merror` element should be in this case. The _core spec_ says `merror` should be treated as `mrow`, but doesn't say how it handles content errors.

I think the merror description in the spec needs a tweak. I've added #66 about that. Basically that tweaks gives a way of describing the error without interfering with the display of the content for this use of `merror`.

With that tweak, @fred-wang's solution of using `mrow` for display (inside of the `merror`) would be compatible with the HTML5 spec and (according to @fred-wang) be easy to implement.

**Comment by bfgeek** [comment]
@davidcarlisle Whatever is decided here, should be specified in the MathML spec, and that removed from the HTML spec.

I've spoken to mozilla engineers about that behaviour and they dislike it, and the technical debt it creates.
cc/ @emilio

**Comment by bfgeek** [comment]
@davidcarlisle Replacing the content with `<merror>` is probably going to be a bunch of work to specify and test that it works correctly, just a few questions that you have to answer:
 - When does this happen within the document lifecycle.
 - Does it operate on the DOM tree, or CSS box tree?
 - Does this update the DOM nodes underneath a `<math>` element, when does this happen?
 - etc.

**Comment by davidcarlisle** [comment]
@bfgeek  I'm perfectly happy to let implementation issues take the lead here, I just wanted to flag the reference in  the HTML5 spec so if we end up specifying something different here that gets pushed back to the w3c and whatwg html editors to update the spec (there are references to the mathml3 W3C CVS draft spec location rather than the currently maintained github drafts which also need updating in the html5 spec).

**Comment by fred-wang** [comment]
@bfgeek I also don't like at all Mozilla / MathML3 / HTMLT5's suggestion to render as if an "merror" was inserted. Firefox actually implements it by displaying a "invalid-markup" message which is not really helpful. The standard way to report an error is via the browser console, which is what Firefox does.

If we want to be consistent with HTML, we should try and render all the content without extra visual indication for error, so so I suggest one of this option:
* https://github.com/mathml-refresh/mathml/issues/15#issuecomment-468584557
* https://github.com/mathml-refresh/mathml/issues/15#issuecomment-469395659

**Comment by fred-wang** [comment]
Just to add more to the discussions, my colleague @rwlbuis pointed out that SVG has some error rules https://www.w3.org/TR/SVG11/implnote.html#ErrorProcessing that implies not displaying some parts of the content. I'm not sure how it is actually implemented in browsers.  

There are also empty elements in MathML like mspace that are not supposed to have any child. I wonder how we should handle them compared to e.g. empty HTML elements. See https://github.com/web-platform-tests/wpt/pull/15722

**Comment by rwlbuis** [comment]
I experimented today with the approach Fred mentioned, using mfrac:
> An alternative option could be to say that invalid markup are just laid out as mrow. 

It turns out to be really easy to do this fallback, basically two lines (pseudo code:)
if (not mfrac.isValid())
  rowLayoutAlgorithm.Layout();

I also tried to add specific code for edge cases (no, one or more than two children of mfrac) but it does not result in nice, clean code.

Finally the tests used for this are up, I expect them to be in a shape to be submitted early next week:
https://github.com/web-platform-tests/wpt/commit/0e30f308141ba8c5e1b1218b6a4797eacc68ecd5

**Comment by fred-wang** [comment]
I don't remember where we discussed that (I think it was during one MathML Core meeting), but basically Igalia's Chromium implementation currently does https://github.com/mathml-refresh/mathml/issues/15#issuecomment-469395659 ; However, @rwlbuis has introduced helper functions that can take a consecutive list of children and laid them out as an mrow. So that would make easier to implement https://github.com/mathml-refresh/mathml/issues/15#issuecomment-468584557 without the pain of having to deal with anonymous box fixup implementation & specification.

Probably, that would not be perfectly equivalent to the valid markup (e.g. there might be differences for CSS selectors for math-style/math-script-level or how embellished operator & implicit mrow are determined for streching). This is not a big deal, because we agree people shouldn't use invalid markup anyway. The goal is just that

1. children are all laid out & rendered.
2. the layout is not altered (i.e. an invalid fraction still looks like a fraction).

I think it was mentioned that this is how Microsoft Word handles invalid markup too.

**Comment by fred-wang** [comment]
Consensus from 2019/09/12 was:
- Send invalid error to the console so that authors know they have to fix it.
- Always render something to the user, at minimum mrow-like layout. Experiment more to see if we can do something that does not alter the visual rendering (Rob already did that for mfrac).

**Comment by NSoiffer** [comment]
From 2020/05/18 meeting:
Rob felt the experiment showed preserving structure and wrapping up the extra elements into the last child was easy to do. The group agreed he should go forward with that and assuming Google is ok with it, expand that "error handling" to the rest of the MathML elements that take a fixed number of arguments as that seems to be in line with the HTML philosophy of trying to do a sensible job of fix up.

There won't be a visual indication of the cleanup, but a warning should be sent to the console.

**Comment by fred-wang** [comment]
Adding the label back because I don't think there is any consensus on the proposed "wrapping up the extra elements" approach and unless I'm missing what @rbuis was referring to, the layout-side experiment he made some months ago (for the mfrac element only) didn't demonstrate at all "it was easy" ; actually it was removed from our branch because it added extra complexity and it is not being upstreamed. There are still open questions like what happens with mmultiscripts, are we ok with "not perfectly equivalent" mentioned above, or if we should consider an anonymous box fixup approach instead (which is probably what is best in the long term) etc

So I don't think the CG is in position to draw any conclusion here without more work on implementations.

On the other hand, the current approach with mrow-like fallback well-described in the text of the spec, implementing it is very easy, chromium reviewers are fine with it and is what the WPT tests assume.

**Comment by fred-wang** [comment]
Adding "need polyfill" as it should be easy to implement in JS what the people have requested to fixup the markup.

**Comment by NSoiffer** [comment]
#175 was resolved with 'render as an mrow'.

I'm not sure what is left, but @bkardell pointed out this can't be closed until there is agreement from the other implementations that rendering as an mrow is ok with them -- they do something different now as pointed out in the initial comment.

**Comment by NSoiffer** [comment]
@fred-wang: I don't think this can be closed unless you heard from other implementers as per @bkardell comment.

**Comment by fred-wang** [comment]
Checking the discussions, it seems Mozilla and Apple are happy with the "render as mrow" approach, but let's see what @bkardell says.
